src/Algebra/Lattice/Stacked.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE DeriveFoldable      #-}
+{-# LANGUAGE DeriveFunctor       #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE DeriveTraversable   #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE Safe                #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators       #-}
+module Algebra.Lattice.Stacked (
+    Stacked(..)
+  , leftOnTopOfRight
+  , rightOnTopOfLeft
+  ) where
+
+import Prelude ()
+import Prelude.Compat
+
+import Algebra.Lattice
+import Algebra.PartialOrd
+
+import Control.DeepSeq       (NFData (..))
+import Control.Monad         (ap)
+import Data.Data             (Data, Typeable)
+import Data.Hashable         (Hashable (..))
+import Data.Universe.Class   (Finite (..), Universe (..))
+import GHC.Generics          (Generic, Generic1)
+
+import qualified Test.QuickCheck as QC
+
+--
+-- Stacked
+--
+
+-- | Stacked two lattices, one on top of another. All minimal elements of upper lattice cover all maximal elements of lower lattice.
+data Stacked a b = Lower a
+              | Upper b
+    deriving ( Eq, Ord, Show, Read, Data, Typeable, Generic, Functor, Foldable, Traversable
+            , Generic1
+            )
+
+-- | Converts @Either@ to @Stacked@ assuming @Either@'s left on top of right
+leftOnTopOfRight :: Either l r -> Stacked r l 
```

Comment:
also here, this is `either Upper Lower` or `either Lower Upper`, which is IMHO better as you don't need to lookup that much.

On other hand having

```haskell
foldStacked :: (l -> r) -> (u -> r) -> Stacked l u -> r
```

would be useful (for the same reason `either` is).

Reply:
Two mentioned function removed, foldStacked function added.